src/zenml/utils/archivable.py

```diff
@@ -81,52 +91,71 @@ def add_directory(self, source: str, destination: str) -> None:
                     self._extra_files[file_destination.as_posix()] = f.read()
 
     def write_archive(
-        self, output_file: IO[bytes], use_gzip: bool = True
+        self,
+        output_file: IO[bytes],
+        archive_type: ArchiveType = ArchiveType.TAR_GZ,
     ) -> None:
         """Writes an archive of the build context to the given file.
 
         Args:
             output_file: The file to write the archive to.
-            use_gzip: Whether to use `gzip` to compress the file.
+            archive_type: The type of archive to create.
         """
         files = self.get_files()
         extra_files = self.get_extra_files()
+        intermediate_fileobj: Optional[Any] = None
```

Comment:
I was merely trying to keep your original code intact and close the `GzipFile` object at the end, but I didn't research if this is actually needed or not. I use `intermediate_fileobj` to make sure I close the `GzipFile` object independently of and in addition to `fileobj`, but only in the `tar.gz` case. It feels a bit awkward, because it's an exception to the overall code flow in this function.

Reply:
Yep we need to create an intermediate fileobj in that case, but this variable could still be used like this instead right?
```python
close_fileobj: bool = False

if archive_type == ArchiveType.TAR_GZ:
  close_fileobj = True
  # set fileobj to the intermediate file here

if close_fileobj:
  fileobj.close()
```

At least to me this is more understandable, but feel free to ignore as well

Reply:
Done